upgrade/upgrade-scripts/rootfs-cleanup

```diff
@@ -0,0 +1,161 @@
+
+    def __le__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "le", version(other.rootfs))
+
+    def __ge__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "ge", version(other.rootfs))
+
+    def __ne__(self, other: object) -> bool:
+        if not isinstance(other, rootfscmp):
+            return NotImplemented
+        return dpkgcmp(version(self.rootfs), "ne", version(other.rootfs))
+
+
+def main() -> None:
+    snapshots = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "2", "-Ho", "name", "-t",
+        "snapshot"
+    ]).decode("UTF-8").splitlines()
+
+    for snap in snapshots:
+        name = snap.split("@")[1]
+
+        #
+        # We filter out any container-delphix snapshots, as those removed by
+        # the calls to "rootfs-container delete", on the relevant filesystems.
+        #
+        if re.search("^container-delphix.\w{7}$", name):
+            snapshots.remove(snap)
+            continue
+
+        #
+        # Out of an abundance of caution, verify the snapshot name matches
+        # our expectations.
+        #
+        if not re.search("^execute-upgrade.\w{7}$", name):
+            raise SystemExit("unexpected snapshot name: '{}'".format(name))
+
+    filesystems = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "1", "-Ho", "name", "-t",
+        "filesystem"
+    ]).decode("UTF-8").splitlines()
+
+    #
+    # We remove the first entry in the list, as that'll refer to the "rpool/ROOT"
+    # dataset, which is not versioned and also not eligible for deletion.
+    #
+    filesystems = filesystems[1:]
+
+    current = subprocess.check_output(["get-appliance-version"
+                                      ]).decode("UTF-8").strip()
+
+    #
+    # Our policy is to keep two delphix versions at any given time; the
+    # currently running version, and the version used prior to the
+    # current one.
+    #
+    # By sorting the list of all filesystems and snapshots on the sytem,
+    # we can easily determine the highest two versions on the system,
+    # and skip those versions (to maintain the policy described above);
+    # this is based on the assumption that the versions of an appliance
+    # will always increase (or remain constant, in the case of rollback)
+    # after an upgrade (i.e. we never upgrade to a lower version).
+    #
+    for rootfs in sorted(filesystems + snapshots, key=rootfscmp)[:-2]:
+        #
+        # In the event of a rollback, we want to be careful to not
+        # delete the currently running version, as well as any versions
+        # that may have failed to upgrade resulting in the rollback
+        # event to begin with. Thus, to account for the exceptional case
+        # of rollback, we simply skip any version that's equal to or
+        # greater than what's currently running.
+        #
```

Comment:
Can you elaborate on what you mean?

You're saying, because we will have multiple rootfs containers/snapshots that will report the same version, and based on how we sort those same-versioned containers/snapshots, without this check we may attempt to remove the one we're currently using?

If so, I think that's valid, and could deserve a comment.

Additionally, I don't think we've discussed how we should handle the cleanup of old hotfix versions.. e.g. in the case that we apply multiple hotfix versions, all based on the same delphix version.. we could potentially accumulate many rootfs snapshots. Based on this logic, they should eventually get cleaned up, when all of them are "old enough", but I'm not sure if that's what we want.

Reply:
> You're saying, because we will have multiple rootfs containers/snapshots that will report the same version, and based on how we sort those same-versioned containers/snapshots, without this check we may attempt to remove the one we're currently using?

Yeah that's right.